alibi_detect/cd/pytorch/lsdd.py

```diff
@@ -0,0 +1,169 @@
+            if self.device.type == 'cpu':
+                print('No GPU detected, fall back on CPU.')
+        else:
+            self.device = torch.device('cpu')
+
+        if self.preprocess_x_ref or self.preprocess_fn is None:
+            x_ref = torch.as_tensor(self.x_ref)
+            self._configure_normalization(x_ref)
+            x_ref = self._normalize(x_ref)
+            self._initialize_kernel(x_ref)
+            self._configure_kernel_centers(x_ref)
+            self.x_ref = x_ref.numpy()
+
+            d = self.x_ref.shape[-1]
+            self.H = GaussianRBF(np.sqrt(2.)*self.kernel.sigma)(self.kernel_centers, self.kernel_centers) * \
+                ((np.pi*self.kernel.sigma**2)**(d/2))  # (Eqn 5)
+
+    def _initialize_kernel(self, x_ref: torch.Tensor):
+        if self.sigma is None:
+            self.kernel = GaussianRBF()
+            _ = self.kernel(x_ref, x_ref, infer_sigma=True)
+        else:
+            sigma = torch.from_numpy(self.sigma)
+            self.kernel = GaussianRBF(sigma)
+
+    def _configure_normalization(self, x_ref: torch.Tensor, eps: float = 1e-7):
+        x_ref_means = x_ref.mean(0)
+        x_ref_stds = x_ref.std(0)
+        self._normalize = lambda x: (x - x_ref_means)/(x_ref_stds + eps)
+
+    def _configure_kernel_centers(self, x_ref: torch.Tensor):
+        "Set aside reference samples to act as kernel centers"
+        perm = torch.randperm(self.x_ref.shape[0])
+        c_inds, non_c_inds = perm[:self.n_kernel_centers], perm[self.n_kernel_centers:]
+        self.kernel_centers = x_ref[c_inds]
+        if np.unique(self.kernel_centers.numpy(), axis=0).shape[0] < self.n_kernel_centers:
+            perturbation = torch.randn(self.kernel_centers.shape)*1e-6
+            self.kernel_centers = self.kernel_centers + perturbation
+        x_ref_eff = x_ref[non_c_inds]  # the effective reference set
+        self.k_xc = self.kernel(x_ref_eff, self.kernel_centers)
+
+    def score(self, x: np.ndarray) -> Tuple[float, float, np.ndarray]:
+        """
+        Compute the p-value resulting from a permutation test using the least-squares density
+        difference as a distance measure between the reference data and the data to be tested.
+
+        Parameters
+        ----------
+        x
+            Batch of instances.
+
+        Returns
+        -------
+        p-value obtained from the permutation test, the LSDD between the reference and test set
+        and the LSDD values from the permutation test.
+        """
+        x_ref, x = self.preprocess(x)
+        x_ref = torch.from_numpy(x_ref).to(self.device)
+        x = torch.from_numpy(x).to(self.device)
+
+        if self.preprocess_fn is not None and self.preprocess_x_ref is False:
+            self._configure_normalization(x_ref)
+            x_ref = self._normalize(x_ref)
+            self._initialize_kernel(x_ref)
+            self._configure_kernel_centers(x_ref)
```

Comment:
I realize this if statement is a bit of an edge case but this would mean every run different kernel centers are configured due to the random permutation?

Reply:
I don't think so? Between runs the `.reset()` method should be called rather than reinitializing the detector.

Reply:
Only if the detector is reinitialised? Even so, would using different kernel centers be a problem?

Reply:
What do you mean with reinitialized? For the offline detector it would be reconfigured every new batch, right?

Reply:
Well it could be used to score multiple batches, particularly once there's saving/loading right?